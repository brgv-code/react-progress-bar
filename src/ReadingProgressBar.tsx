import React, { useEffect, useState, CSSProperties } from 'react'

export interface ReadingProgressBarProps {
	/**
	 * Base color for the progress bar
	 */
	color?: string
	/**
	 * Height of the progress bar in pixels
	 */
	height?: number
	/**
	 * Additional CSS classes
	 */
	className?: string
	/**
	 * Whether to use a gradient effect
	 */
	useGradient?: boolean
	/**
	 * Whether to use a glow effect
	 */
	useGlow?: boolean
	/**
	 * Custom gradient colors (requires useGradient=true)
	 * Format: Array of colors for the gradient
	 */
	gradientColors?: string[]
	/**
	 * Intensity of the glow effect (0-1)
	 */
	glowIntensity?: number
	/**
	 * Minimum scroll position (in pixels) before showing the progress bar
	 */
	showAfterScroll?: number
	/**
	 * Position of the progress bar
	 */
	position?: 'top' | 'bottom'
	/**
	 * Z-index of the progress bar
	 */
	zIndex?: number
	/**
	 * Whether to show the progress bar initially
	 */
	initiallyVisible?: boolean
	/**
	 * Custom styles to apply to the progress bar
	 */
	style?: CSSProperties
}

/**
 * A reading progress bar component that shows how far the user has scrolled down the page.
 *
 * @example
 * // Basic usage
 * <ReadingProgressBar />
 *
 * @example
 * // Custom styling
 * <ReadingProgressBar
 *   color="#FF5500"
 *   height={5}
 *   useGradient={true}
 *   useGlow={true}
 *   position="bottom"
 * />
 */
export const ReadingProgressBar: React.FC<ReadingProgressBarProps> = ({
	color = '#FFDD00',
	height = 3,
	className = '',
	useGradient = true,
	useGlow = true,
	gradientColors,
	glowIntensity = 0.7,
	showAfterScroll = 100,
	position = 'top',
	zIndex = 40,
	initiallyVisible = false,
	style = {},
}) => {
	const [readingProgress, setReadingProgress] = useState(0)
	const [isVisible, setIsVisible] = useState(initiallyVisible)

	useEffect(() => {
		const updateReadingProgress = () => {
			const currentPosition = window.scrollY
			const scrollHeight = document.body.scrollHeight - window.innerHeight

			// Show/hide based on scroll position
			setIsVisible(currentPosition > showAfterScroll)

			// Calculate reading progress
			if (scrollHeight) {
				setReadingProgress(Number((currentPosition / scrollHeight).toFixed(2)))
			}
		}

		// Update on mount
		updateReadingProgress()

		// Add scroll event listener
		window.addEventListener('scroll', updateReadingProgress)

		// Clean up
		return () => window.removeEventListener('scroll', updateReadingProgress)
	}, [showAfterScroll])

	// Default gradient colors if not provided
	const defaultGradientColors = [color, '#FFA500', '#FF8C00']
	const finalGradientColors = gradientColors || defaultGradientColors

	// Create the gradient string
	const gradientString = `linear-gradient(90deg, ${finalGradientColors.join(', ')})`

	// Calculate glow with proper opacity
	const hexOpacity = Math.floor(glowIntensity * 255)
		.toString(16)
		.padStart(2, '0')

	// Set custom CSS variables for the progress bar
	const customStyle = {
		'--progress-color': color,
		'--progress-gradient': useGradient ? gradientString : color,
		'--progress-glow': useGlow ? `0 0 10px ${color}${hexOpacity}` : 'none',
	} as React.CSSProperties

	return (
		<progress
			className={`reading-progress ${className} ${isVisible ? 'visible' : 'hidden'}`}
			value={readingProgress}
			max='1'
			style={{
				position: 'sticky',
				[position]: 0,
				zIndex,
				width: '100%',
				height: `${height}px`,
				border: 'none',
				appearance: 'none',
				backgroundColor: 'transparent',
				transition: 'opacity 300ms ease-in-out',
				...customStyle,
				...style,
			}}
			aria-label='Reading progress'
		/>
	)
}

export default ReadingProgressBar
