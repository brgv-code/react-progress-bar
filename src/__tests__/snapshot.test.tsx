import React from 'react';
import { render } from '@testing-library/react';
import ReadingProgressBar from '../ReadingProgressBar';

Object.defineProperty(window, 'scrollY', { value: 0, writable: true });
Object.defineProperty(document.body, 'scrollHeight', {
  value: 1000,
  writable: true,
});
Object.defineProperty(window, 'innerHeight', { value: 500, writable: true });

describe('ReadingProgressBar Snapshots', () => {
  it('matches snapshot with default props', () => {
    const { container } = render(<ReadingProgressBar />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with custom styling', () => {
    const { container } = render(
      <ReadingProgressBar
        color="#FF5500"
        height={8}
        useGradient={true}
        useGlow={true}
        position="bottom" 
        zIndex={100}
        initiallyVisible={true}
        className="custom-progress"
        style={{ borderRadius: '4px' }}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with gradient disabled', () => {
    const { container } = render(
      <ReadingProgressBar color="#3498db" useGradient={false} useGlow={true} />
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with glow disabled', () => {
    const { container } = render(
      <ReadingProgressBar color="#e74c3c" useGradient={true} useGlow={false} />
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with custom gradient colors', () => {
    const { container } = render(
      <ReadingProgressBar useGradient={true} gradientColors={['#3498db', '#2ecc71', '#f1c40f']} />
    );
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with all effects disabled', () => {
    const { container } = render(
      <ReadingProgressBar color="#9b59b6" useGradient={false} useGlow={false} />
    );
    expect(container).toMatchSnapshot();
  });
});
