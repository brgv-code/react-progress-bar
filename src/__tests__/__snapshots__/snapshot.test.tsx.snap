// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadingProgressBar Snapshots matches snapshot with all effects disabled 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress  hidden"
    max="1"
    style="position: sticky; top: 0px; z-index: 40; width: 100%; height: 3px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #9b59b6; --progress-gradient: #9b59b6; --progress-glow: none;"
    value="0"
  />
</div>
`;

exports[`ReadingProgressBar Snapshots matches snapshot with custom gradient colors 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress  hidden"
    max="1"
    style="position: sticky; top: 0px; z-index: 40; width: 100%; height: 3px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #FFDD00; --progress-gradient: linear-gradient(90deg, #3498db, #2ecc71, #f1c40f); --progress-glow: 0 0 10px #FFDD00b2;"
    value="0"
  />
</div>
`;

exports[`ReadingProgressBar Snapshots matches snapshot with custom styling 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress custom-progress hidden"
    max="1"
    style="position: sticky; bottom: 0px; z-index: 100; width: 100%; height: 8px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #FF5500; --progress-gradient: linear-gradient(90deg, #FF5500, #FFA500, #FF8C00); --progress-glow: 0 0 10px #FF5500b2; border-radius: 4px;"
    value="0"
  />
</div>
`;

exports[`ReadingProgressBar Snapshots matches snapshot with default props 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress  hidden"
    max="1"
    style="position: sticky; top: 0px; z-index: 40; width: 100%; height: 3px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #FFDD00; --progress-gradient: linear-gradient(90deg, #FFDD00, #FFA500, #FF8C00); --progress-glow: 0 0 10px #FFDD00b2;"
    value="0"
  />
</div>
`;

exports[`ReadingProgressBar Snapshots matches snapshot with glow disabled 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress  hidden"
    max="1"
    style="position: sticky; top: 0px; z-index: 40; width: 100%; height: 3px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #e74c3c; --progress-gradient: linear-gradient(90deg, #e74c3c, #FFA500, #FF8C00); --progress-glow: none;"
    value="0"
  />
</div>
`;

exports[`ReadingProgressBar Snapshots matches snapshot with gradient disabled 1`] = `
<div>
  <progress
    aria-label="Reading progress"
    class="reading-progress  hidden"
    max="1"
    style="position: sticky; top: 0px; z-index: 40; width: 100%; height: 3px; appearance: none; background-color: transparent; transition: opacity 300ms ease-in-out; --progress-color: #3498db; --progress-gradient: #3498db; --progress-glow: 0 0 10px #3498dbb2;"
    value="0"
  />
</div>
`;
